### PR TITLE
Misc Cleanup

### PR DESCRIPTION
**Pull request by cgzones**
Mark several structure members const.
E.g. settings should only be changed by the Settings Screen and not through a Process pointer.

(Dropped the hashtable changes)

### CODE REVIEW

File: ProcessList.c [hunk 1]
```diff
@@ -87,7 +87,7 @@ void ProcessList_add(ProcessList* this, Process* p) {
 
    assert(Vector_indexOf(this->processes, p, Process_pidCompare) != -1);
    assert(Hashtable_get(this->processTable, p->pid) != NULL);
-   assert(Hashtable_count(this->processTable) == Vector_count(this->processes));
+   assert(Hashtable_count(this->processTable) == (unsigned int)Vector_count(this->processes));
```
Comment (BenBE) [inline ProcessList.c:90]: Shouldn't be the types for amounts in `*_count` functions be compatible?

File: ProcessList.c [hunk 2]
```diff
@@ -100,7 +100,7 @@ void ProcessList_remove(ProcessList* this, Process* p) {
    assert(idx != -1);
    if (idx >= 0) Vector_remove(this->processes, idx);
    assert(Hashtable_get(this->processTable, pid) == NULL); (void)pid;
-   assert(Hashtable_count(this->processTable) == Vector_count(this->processes));
+   assert(Hashtable_count(this->processTable) == (unsigned int)Vector_count(this->processes));
```
Comment (BenBE) [inline ProcessList.c:103]: Dito.

File: ProcessList.c [hunk 3]
```diff
@@ -142,20 +142,21 @@ static void ProcessList_buildTree(ProcessList* this, pid_t pid, int level, int i
...
+static long ProcessList_treeProcessCompare(const void* v1, const void* v2) {
+   const Process *p1 = (const Process*)v1;
+   const Process *p2 = (const Process*)v2;
+
+   return p1->pid - p2->pid;
+}
+
 void ProcessList_sort(ProcessList* this) {
    if (!this->settings->treeView) {
       Vector_insertionSort(this->processes);
```
Comment (BenBE) [inline ProcessList.c:154]: Couldn't `Vector_quickSortCustomCompare` be also used for normal sorting (given the right comparison function)?

File: darwin/DarwinProcess.c [hunk 4]
```diff
@@ -37,7 +37,7 @@ DarwinProcess* DarwinProcess_new(Settings* settings) {
    this->stime = 0;
    this->taskAccess = true;
 
-   return this;
+   return &this->super;
```
Comment (BenBE) [inline darwin/DarwinProcess.c:40]: Use some central `AsProcess` magic that enforces `&(*this)` and `&this->super` to be the same pointers (mod type).

**Review by BenBE — COMMENTED** [review]
Do we have some place for central (static) checking of type constraints? E.g. with this change we really require the .super field to be at offset 0 in the struct. While this is true for all current platforms I'd feel better if this was checked somewhere (statically).

**Comment by BenBE** [comment]
Any reason to copy `settings` over to each process instead of having each process just know which process list it belongs to and reference from there?

File: ProcessList.c (hunk 3 above)
Comment (cgzones) [reply ProcessList.c:154]: `Vector_quickSort` uses the *right* one?

File: darwin/DarwinProcess.c (hunk 4 above)
Comment (cgzones) [reply darwin/DarwinProcess.c:40]: All the Object pseudo inheritance bases on the fact that a struct address equals the address of the first field.

**Comment by cgzones** [comment]
> Any reason to copy settings over to each process instead of having each process just know which process list it belongs to and reference from there?

What's the difference of storing the ProcessList instead of Settings?

**Comment by BenBE** [comment]
> > Any reason to copy settings over to each process instead of having each process just know which process list it belongs to and reference from there?
> 
> What's the difference of storing the ProcessList instead of Settings?

With ProcessList* you can store more things in a central place.
The ProcessList is added to Process for #240 too.

File: ProcessList.c (hunk 3 above)
Comment (BenBE) [reply ProcessList.c:154]: Just wondered if we have code duplication there.

File: Vector.c [hunk 5]
```diff
@@ -54,13 +54,13 @@ static bool Vector_isConsistent(const Vector* this) {
...
-int Vector_count(const Vector* this) {
-   int items = 0;
+unsigned int Vector_count(const Vector* this) {
+   unsigned int items = 0;
    for (int i = 0; i < this->items; i++) {
       if (this->array[i])
          items++;
    }
-   assert(items == this->items);
+   assert(items == (unsigned int)this->items);
```
Comment (BenBE) [inline Vector.c:63]: Chance to avoid this cast too?

Comment (cgzones) [reply Vector.c:63]: That will cascade into many `int` -> `unsigned int` changes throughout the code; maybe at some point

File: darwin/DarwinProcess.c (hunk 4 above)
Comment (cgzones) [reply darwin/DarwinProcess.c:40]: It's guaranteed by the C Standard: 

6.7.2.1-13 states:
> Apointer to astructure  object,  suitably  converted,  points  to  its  initial  member  (or  if  that  member  is  abit-field,  then  to  the  unit  in  which  it  resides),  and  vice  versa.   Theremay  be  unnamedpadding within a structure object, but not at its beginning.

http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf page 115

**Pull request merged**